File: scripts/build_super_regions.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const turf = require('@turf/turf');
const { parse } = require('csv-parse/sync');

function normalizeLabel(value) {
  const text = String(value ?? '').trim();
  return text
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '');
}

function parseArgs(argv) {
  const args = {};
  for (let i = 0; i < argv.length; i += 1) {
    const token = argv[i];
    if (!token.startsWith('--')) {
      continue;
    }

    const key = token.slice(2);
    const next = argv[i + 1];

    if (!next || next.startsWith('--')) {
      args[key] = true;
      continue;
    }

    args[key] = next;
    i += 1;
  }
  return args;
}

function printHelp() {
  console.log(`Usage:
  node scripts/build_super_regions.js --input <file.geojson> --mapping <file.json|file.csv> --output <file.geojson> [options]

Options:
  --source-field <name>            Input property field used for mapping keys (default: NAME_1)
  --target-field <name>            Output group field (default: macro_region)
  --mapping-source-column <name>   Source column for CSV/JSON-list mapping (default: source)
  --mapping-target-column <name>   Target column for CSV/JSON-list mapping (default: target)
  --normalize                      Normalize labels before matching
  --on-missing <mode>              error | drop | keep-source (default: keep-source)
  --help                           Show this help message`);
}

function readJsonFile(filePath) {
  return JSON.parse(fs.readFileSync(filePath, 'utf8'));
}

function loadCsvMapping(filePath, sourceColumn, targetColumn) {
  const raw = fs.readFileSync(filePath, 'utf8');
  const records = parse(raw, {
    columns: true,
    skip_empty_lines: true,
    trim: true
  });

  if (!Array.isArray(records) || records.length === 0) {
    throw new Error('CSV mapping is empty.');
  }

  const sample = records[0];
  if (!(sourceColumn in sample) || !(targetColumn in sample)) {
    throw new Error(`CSV mapping must contain columns '${sourceColumn}' and '${targetColumn}'.`);
  }

  const mapping = {};
  for (const row of records) {
    const source = String(row[sourceColumn] ?? '').trim();
    const target = String(row[targetColumn] ?? '').trim();
    if (!source) {
      continue;
    }
    mapping[source] = target;
  }

  if (Object.keys(mapping).length === 0) {
    throw new Error('Mapping file is empty.');
  }

  return mapping;
}

function loadMapping(filePath, sourceColumn, targetColumn, useNormalization) {
  if (!fs.existsSync(filePath)) {
    throw new Error(`Mapping file not found: ${filePath}`);
  }

  const ext = path.extname(filePath).toLowerCase();
  let rawMapping = {};

  if (ext === '.json') {
    const data = readJsonFile(filePath);
    if (Array.isArray(data)) {
      for (const row of data) {
        if (!row || typeof row !== 'object') {
          throw new Error('JSON list mapping must contain objects.');
        }
        if (!(sourceColumn in row) || !(targetColumn in row)) {
          throw new Error(`JSON list mapping rows must contain '${sourceColumn}' and '${targetColumn}'.`);
        }
        rawMapping[String(row[sourceColumn])] = String(row[targetColumn]);
      }
    } else if (data && typeof data === 'object') {
      const entries = Object.entries(data);
      const usesCompactGroups = entries.some(([, value]) => Array.isArray(value));

      if (usesCompactGroups) {
        // Compact format:
        // {
        //   "North": ["ChiangMai", "ChiangRai"],
        //   "Central": ["Bangkok"]
        // }
        for (const [targetGroup, sourceList] of entries) {
          if (!Array.isArray(sourceList)) {
            throw new Error(
              'Mixed JSON mapping object is not supported. Use either {"source":"target"} ' +
              'or {"target":["source1","source2"]}.'
            );
          }

          for (const sourceName of sourceList) {
            const sourceKey = String(sourceName);
            if (sourceKey in rawMapping && rawMapping[sourceKey] !== String(targetGroup)) {
              throw new Error(
                `Conflicting JSON group mapping for source '${sourceKey}': ` +
                `'${rawMapping[sourceKey]}' vs '${targetGroup}'.`
              );
            }
            rawMapping[sourceKey] = String(targetGroup);
          }
        }
      } else {
        rawMapping = Object.fromEntries(
          entries.map(([k, v]) => [String(k), String(v)])
        );
      }
    } else {
      throw new Error('Unsupported JSON mapping format.');
    }
  } else if (ext === '.csv') {
    rawMapping = loadCsvMapping(filePath, sourceColumn, targetColumn);
  } else {
    throw new Error('Mapping file must be .json or .csv');
  }

  if (Object.keys(rawMapping).length === 0) {
    throw new Error('Mapping file is empty.');
  }

  if (!useNormalization) {
    return rawMapping;
  }

  const normalized = {};
  for (const [key, value] of Object.entries(rawMapping)) {
    const normKey = normalizeLabel(key);
    if (normKey in normalized && normalized[normKey] !== value) {
      throw new Error(`Conflicting mapping after normalization for key '${key}' -> '${normKey}'.`);
    }
    normalized[normKey] = value;
  }

  return normalized;
}

function ensureFeatureCollection(data) {
  if (!data || data.type !== 'FeatureCollection' || !Array.isArray(data.features)) {
    throw new Error('Input GeoJSON must be a FeatureCollection.');
  }
}

function dissolveGroup(features) {
  const sourceArea = features.reduce((sum, feature) => {
    try {
      return sum + turf.area(feature);
    } catch (err) {
      return sum;
    }
  }, 0);

  const isReasonableGeometry = (geometry) => {
    try {
      const outputArea = turf.area(turf.feature(geometry));
      if (!Number.isFinite(outputArea) || outputArea <= 0) {
        return false;
      }

      if (sourceArea > 0 && outputArea > sourceArea * 1.35) {
        return false;
      }

      return true;
    } catch (err) {
      return false;
    }
  };

  if (features.length === 1) {
    return features[0].geometry;
  }

  try {
    const dissolved = turf.dissolve(turf.featureCollection(features));
    if (dissolved && Array.isArray(dissolved.features) && dissolved.features[0]?.geometry) {
      const geometry = dissolved.features[0].geometry;
      if (isReasonableGeometry(geometry)) {
        return geometry;
      }
    }
  } catch (err) {
    // fallback below
  }

  let merged = turf.feature(features[0].geometry);
  for (let i = 1; i < features.length; i += 1) {
    const next = turf.feature(features[i].geometry);
    let unioned = null;

    try {
      unioned = turf.union(merged, next);
    } catch (err) {
      unioned = null;
    }

    if (!unioned) {
      const combined = turf.combine(turf.featureCollection([merged, next]));
      unioned = combined.features[0];
    }

    merged = unioned;
  }

  if (merged?.geometry && isReasonableGeometry(merged.geometry)) {
    return merged.geometry;
  }

  const combined = turf.combine(turf.featureCollection(features));
  if (combined?.features?.[0]?.geometry) {
    return combined.features[0].geometry;
  }

  return features[0].geometry;
}

function main() {
  const args = parseArgs(process.argv.slice(2));

  if (args.help) {
    printHelp();
    process.exit(0);
  }

  const inputPath = args.input;
  const mappingPath = args.mapping;
  const outputPath = args.output;

  if (!inputPath || !mappingPath || !outputPath) {
    printHelp();
    throw new Error('Missing required arguments: --input, --mapping, --output');
  }

  const sourceField = args['source-field'] || 'NAME_1';
  const targetField = args['target-field'] || 'macro_region';
  const mappingSourceColumn = args['mapping-source-column'] || 'source';
  const mappingTargetColumn = args['mapping-target-column'] || 'target';
  const useNormalization = Boolean(args.normalize);
  const onMissing = args['on-missing'] || 'keep-source';

  if (!['error', 'drop', 'keep-source'].includes(onMissing)) {
    throw new Error("--on-missing must be one of: error, drop, keep-source");
  }

  if (!fs.existsSync(inputPath)) {
    throw new Error(`Input file not found: ${inputPath}`);
  }

  const mapping = loadMapping(mappingPath, mappingSourceColumn, mappingTargetColumn, useNormalization);
  const geojson = readJsonFile(inputPath);
  ensureFeatureCollection(geojson);

  const missingValues = new Set();
  const selectedFeatures = [];

  for (const feature of geojson.features) {
    const properties = feature && feature.properties ? feature.properties : {};
    const sourceValue = properties[sourceField];

    if (sourceValue === undefined || sourceValue === null) {
      missingValues.add(String(sourceValue));
      continue;
    }

    const sourceString = String(sourceValue);
    const sourceKey = useNormalization ? normalizeLabel(sourceString) : sourceString;
    let targetValue = mapping[sourceKey];

    if (targetValue === undefined || targetValue === null || targetValue === '') {
      missingValues.add(sourceString);

      if (onMissing === 'drop') {
        continue;
      }
      if (onMissing === 'keep-source') {
        targetValue = sourceString;
      }
    }

    if (targetValue === undefined || targetValue === null || targetValue === '') {
      continue;
    }

    selectedFeatures.push({
      type: 'Feature',
      properties: {
        ...properties,
        [targetField]: String(targetValue)
      },
      geometry: feature.geometry
    });
  }

  const missingList = Array.from(missingValues).sort();
  if (missingList.length > 0 && onMissing === 'error') {
    const preview = missingList.slice(0, 20).join(', ');
    const suffix = missingList.length > 20 ? ` ... (+${missingList.length - 20} more)` : '';
    throw new Error(
      `Some regions are missing in mapping. Count=${missingList.length}. Examples: ${preview}${suffix}. ` +
      'Use --on-missing drop or --on-missing keep-source if desired.'
    );
  }

  if (selectedFeatures.length === 0) {
    throw new Error('No features left to export after filtering missing mappings.');
  }

  const grouped = new Map();
  for (const feature of selectedFeatures) {
    const group = feature.properties[targetField];
    if (!grouped.has(group)) {
      grouped.set(group, []);
    }
    grouped.get(group).push(feature);
  }

  const outputFeatures = [];
  const sortedGroups = Array.from(grouped.keys()).sort((a, b) => String(a).localeCompare(String(b)));

  for (const group of sortedGroups) {
    const groupFeatures = grouped.get(group);
    const geometry = dissolveGroup(groupFeatures);
    outputFeatures.push({
      type: 'Feature',
      properties: {
        [targetField]: group
      },
      geometry
    });
  }

  const result = {
    type: 'FeatureCollection',
    features: outputFeatures
  };
  if (geojson.crs) {
    result.crs = geojson.crs;
  }

  fs.mkdirSync(path.dirname(outputPath), { recursive: true });
  fs.writeFileSync(outputPath, JSON.stringify(result, null, 2));

  console.log(`Input features: ${geojson.features.length}`);
  console.log(`Output super-regions: ${outputFeatures.length}`);
  console.log(`Missing source regions: ${missingList.length}`);
  if (missingList.length > 0) {
    console.log(`Missing examples: ${missingList.slice(0, 20).join(', ')}`);
  }
  console.log(`Output written to: ${outputPath}`);
}

try {
  main();
} catch (err) {
  console.error(`Error: ${err.message}`);
  process.exit(1);
}

File: scripts/build_super_regions.py
#!/usr/bin/env python3
"""Build super-regions from a GeoJSON layer and a mapping file.

Examples:
  python3 scripts/build_super_regions.py \
    --input Africa/gadm41_MAR_1.json \
    --mapping mappings/morocco_macro_regions.json \
    --output output/morocco_macro_regions.geojson
"""

from __future__ import annotations

import argparse
import csv
import json
import re
import sys
import unicodedata
from pathlib import Path
from typing import Any


class CliError(Exception):
    """Raised when user input is invalid."""


def normalize_label(value: Any) -> str:
    """Normalize labels to match strings with accents/spaces/punctuation differences."""
    text = str(value or "").strip()
    text = unicodedata.normalize("NFKD", text)
    text = "".join(ch for ch in text if not unicodedata.combining(ch))
    text = text.lower()
    text = re.sub(r"[^a-z0-9]+", "", text)
    return text


def load_mapping(
    path: Path,
    mapping_source_column: str,
    mapping_target_column: str,
    use_normalization: bool,
) -> dict[str, str]:
    """Load mapping from JSON or CSV.

    JSON accepted shapes:
      - {"ProvinceA": "North", "ProvinceB": "South"}
      - [{"source": "ProvinceA", "target": "North"}, ...]

    CSV expected columns by default: source,target
    """
    if not path.exists():
        raise CliError(f"Mapping file not found: {path}")

    suffix = path.suffix.lower()
    raw_mapping: dict[str, str]

    if suffix == ".json":
        data = json.loads(path.read_text(encoding="utf-8"))
        if isinstance(data, dict):
            raw_mapping = {str(k): str(v) for k, v in data.items()}
        elif isinstance(data, list):
            raw_mapping = {}
            for row in data:
                if not isinstance(row, dict):
                    raise CliError("JSON list mapping must contain objects.")
                if mapping_source_column not in row or mapping_target_column not in row:
                    raise CliError(
                        "JSON list mapping rows must contain "
                        f"'{mapping_source_column}' and '{mapping_target_column}'."
                    )
                raw_mapping[str(row[mapping_source_column])] = str(row[mapping_target_column])
        else:
            raise CliError("Unsupported JSON mapping format.")

    elif suffix == ".csv":
        raw_mapping = {}
        with path.open("r", encoding="utf-8", newline="") as fh:
            reader = csv.DictReader(fh)
            if not reader.fieldnames:
                raise CliError("CSV mapping is empty.")
            if mapping_source_column not in reader.fieldnames or mapping_target_column not in reader.fieldnames:
                raise CliError(
                    f"CSV mapping must contain columns '{mapping_source_column}' and '{mapping_target_column}'."
                )
            for row in reader:
                raw_mapping[str(row[mapping_source_column])] = str(row[mapping_target_column])
    else:
        raise CliError("Mapping file must be .json or .csv")

    if not raw_mapping:
        raise CliError("Mapping file is empty.")

    if use_normalization:
        normalized = {}
        for key, value in raw_mapping.items():
            norm_key = normalize_label(key)
            if norm_key in normalized and normalized[norm_key] != value:
                raise CliError(
                    f"Conflicting mapping after normalization for key '{key}' -> '{norm_key}'."
                )
            normalized[norm_key] = value
        return normalized

    return raw_mapping


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Merge GeoJSON regions into super-regions using a mapping file.",
    )
    parser.add_argument("--input", required=True, help="Input GeoJSON path (regions/provinces).")
    parser.add_argument("--mapping", required=True, help="Mapping file (.json or .csv).")
    parser.add_argument("--output", required=True, help="Output GeoJSON path.")

    parser.add_argument(
        "--source-field",
        default="NAME_1",
        help="Property name in input GeoJSON used to match mapping keys (default: NAME_1).",
    )
    parser.add_argument(
        "--target-field",
        default="macro_region",
        help="Output property name for grouped region label (default: macro_region).",
    )

    parser.add_argument(
        "--mapping-source-column",
        default="source",
        help="Source column/key name for CSV or JSON-list mapping files (default: source).",
    )
    parser.add_argument(
        "--mapping-target-column",
        default="target",
        help="Target column/key name for CSV or JSON-list mapping files (default: target).",
    )

    parser.add_argument(
        "--normalize",
        action="store_true",
        help=(
            "Normalize labels before matching (remove accents/spaces/punctuation + lowercase). "
            "Useful when mapping names differ slightly."
        ),
    )

    parser.add_argument(
        "--on-missing",
        choices=["error", "drop", "keep-source"],
        default="error",
        help=(
            "What to do when source regions are missing in mapping: "
            "error (default), drop, or keep-source."
        ),
    )

    parser.add_argument(
        "--fix-invalid",
        action="store_true",
        help="Try to repair invalid geometries before dissolve using a zero-width buffer.",
    )

    return parser.parse_args()


def main() -> int:
    args = parse_args()

    try:
        import geopandas as gpd
    except ImportError as exc:
        print(
            "Missing dependency: geopandas.\n"
            "Install it with one of these commands:\n"
            "  pip install geopandas\n"
            "  or\n"
            "  uv pip install geopandas",
            file=sys.stderr,
        )
        raise SystemExit(1) from exc

    input_path = Path(args.input)
    output_path = Path(args.output)

    if not input_path.exists():
        raise CliError(f"Input file not found: {input_path}")

    mapping = load_mapping(
        Path(args.mapping),
        mapping_source_column=args.mapping_source_column,
        mapping_target_column=args.mapping_target_column,
        use_normalization=args.normalize,
    )

    gdf = gpd.read_file(input_path)

    if args.source_field not in gdf.columns:
        raise CliError(
            f"Field '{args.source_field}' not found in input. Available fields: {', '.join(gdf.columns)}"
        )

    source_series = gdf[args.source_field].astype(str)

    if args.normalize:
        source_keys = source_series.map(normalize_label)
    else:
        source_keys = source_series

    gdf[args.target_field] = source_keys.map(mapping)

    missing_mask = gdf[args.target_field].isna()
    missing_values = sorted(gdf.loc[missing_mask, args.source_field].astype(str).unique().tolist())

    if missing_values and args.on_missing == "error":
        preview = ", ".join(missing_values[:20])
        suffix = "" if len(missing_values) <= 20 else f" ... (+{len(missing_values) - 20} more)"
        raise CliError(
            "Some regions are missing in mapping. "
            f"Count={len(missing_values)}. Examples: {preview}{suffix}. "
            "Use --on-missing drop or --on-missing keep-source if desired."
        )

    if missing_values and args.on_missing == "drop":
        gdf = gdf.loc[~missing_mask].copy()

    if missing_values and args.on_missing == "keep-source":
        gdf.loc[missing_mask, args.target_field] = gdf.loc[missing_mask, args.source_field].astype(str)

    if gdf.empty:
        raise CliError("No features left to export after filtering missing mappings.")

    if args.fix_invalid:
        gdf["geometry"] = gdf.geometry.buffer(0)

    dissolved = gdf.dissolve(by=args.target_field, as_index=False)
    dissolved = dissolved[[args.target_field, "geometry"]].sort_values(by=args.target_field).reset_index(drop=True)

    output_path.parent.mkdir(parents=True, exist_ok=True)
    dissolved.to_file(output_path, driver="GeoJSON")

    print(f"Input features: {len(source_series)}")
    print(f"Output super-regions: {len(dissolved)}")
    print(f"Missing source regions: {len(missing_values)}")
    if missing_values:
        print("Missing examples:", ", ".join(missing_values[:20]))
    print(f"Output written to: {output_path}")

    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except CliError as err:
        print(f"Error: {err}", file=sys.stderr)
        raise SystemExit(1)
